README.md
OmniMarkupPreviewer
===================

Description
-----------

OmniMarkupPreviewer is a plugin for both [Sublime Text 2] and [Sublime Text 3]
that preview markups in web browsers. OmniMarkupPreviewer renders markups into
htmls and send it to web browser in the backgound, which enables a live preview.
Besides, OmniMarkupPreviewer provide support for exporting result to html file 
as well.

[Sublime Text 2]: http://www.sublimetext.com/2
[Sublime Text 3]: http://www.sublimetext.com/3

OmniMarkupPreviewer has builtin support following markups:

* [Markdown](http://daringfireball.net/projects/markdown/)
* [reStructuredText](http://docutils.sourceforge.net/rst.html)
* [WikiCreole](http://wikicreole.org/)
* [Textile](http://www.textism.com/tools/textile/)
* [Pod](http://search.cpan.org/dist/perl/pod/perlpod.pod) (Requires Perl >= `5.10`
  and can be found in `PATH`, if the perl version < `5.10`, `Pod::Simple` should be
  installed from `CPAN`.)
* [RDoc](http://rdoc.sourceforge.net/) (Requires `ruby` in your `PATH`)
* [Org Mode](http://orgmode.org) (Requires `ruby`, and gem `org-ruby` should be installed)
* [MediaWiki](http://www.mediawiki.org/) (Requires `ruby`, as well as gem `wikicloth`)
* [AsciiDoc](http://www.methods.co.nz/asciidoc/) (Requires `ruby`, as well as gem `asciidoctor`)
* Literate Haskell


### Donation

If you find my work useful, please consider buying me a cup of coffee, all
donations are much appreciated :)

[![Donate via PayPal](https://dl.dropbox.com/u/2451120/donate-with-paypal.png)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=TD4HF4X2R53NE)

Installation
------------

### With the Package Control plugin
The easiest way to install OmniMarkupPreviewer is through [Package Control].

[Package Control]: http://wbond.net/sublime_packages/package_control

Once you have Package Control installed, restart Sublime Text.

1. Bring up the Command Palette (<kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>P</kbd>
   on Windows and Linux. <kbd>⌘</kbd>+<kbd>⇧</kbd>+<kbd>P</kbd> on OS X).
2. Type "Install" and select "Package Control: Install Package".
3. Select "OmniMarkupPreviewer" from list.

The advantage of using Package Control is that it will keep OmniMarkupPreviewer
up to date automatically.


### Manual Install
**Without Git:**
[Download](https://github.com/timonwong/OmniMarkupPreviewer) the latest source
code, and extract to the Packages directory.

**With Git:**
Type the following command in your Sublime Text Packages directory:

`git clone git://github.com/timonwong/OmniMarkupPreviewer.git`

The "Packages" directory for Sublime Text 2 is located at:

* **Windows:**  `%APPDATA%\Sublime Text 2\Packages\`
* **Linux:**    `~/.config/sublime-text-2/Packages/`
* **OS X:**     `~/Library/Application Support/Sublime Text 2/Packages/`

And for the Sublime Text 3:

* **Windows:**  `%APPDATA%\Sublime Text 3\Packages\`
* **Linux:**    `~/.config/sublime-text-3/Packages/`
* **OS X:**     `~/Library/Application Support/Sublime Text 3/Packages/`


Usage
-----

### Key Bindings

The default key bindings:

**Windows, Linux:**

* <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>O</kbd>: Preview Markup in Browser.
* <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>X</kbd>: Export Markup as HTML.
* <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>C</kbd>: Copy Markup as HTML.

**OSX:**

* <kbd>⌘</kbd>+<kbd>⌥</kbd>+<kbd>O</kbd>: Preview Markup in Browser.
* <kbd>⌘</kbd>+<kbd>⌥</kbd>+<kbd>X</kbd>: Export Markup as HTML.
* <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>C</kbd>: Copy Markup as HTML.

### Command Palette

Available OmniMarkupPreviewer commands in the command palette:

* `OmniMarkupPreviewer: Preview Current Markup in Browser`
* `OmniMarkupPreviewer: Export Current Markup as HTML`
* `OmniMarkupPreviewer: Empty Cache`


**NOTE** Command will not be available (greyed out) if current file syntax 
(or filename) doesn't match any of OmniMarkupPreviewer renderers.


What's New
----------

**For a complete list of changes, see [CHANGELOG.md](./CHANGELOG.md)**

**v3.0.0 (02/28/2015)**

* Use new flat github style (old style renamed to github-v1).
* Upgrade pygments library to v2.0.2, with more lexers.
* Upgrade `Python-Markdown` package to v2.4.1 final.
* Add `.mmd` file extension support (MultiMarkdown).
* Upgrade jQuery to 2.1.3 (Say good bye to IE8).
* `server_host` setting is now taken into account when launching preview in browser.

**v2.0.9 (08/02/2014)**

* Add syntax highlighting support for reStructureText.
* Prefer to use `xdg-open` under linux environments.

**v2.0.8 (04/15/2014)**

* Fix Sublime Text 2 compatibility (crash).

**v2.0.7 (04/14/2014)**

* Fix unreliable default setting overriding, which confuses a lot of users when
  customizing their settings.
* Prevent preview from the console, which will simply lead to crash.

**v2.0.6 (01/27/2014)**

* Fix Sublime Text 2 compatibility (markdown).

**v2.0.5 (01/24/2014)**

* MathJax library is now bundled directly instead of on demand downloading.
* Python-Markdown module now upgraded to v2.4.
* Fix incompatibility when using Sublime Text 3 under XFCE.

**v2.0.4 (08/10/2013)**

* Fix ruby gems loading in OSX (Required by RDoc, AsciiDoc, Org Mode and MediaWiki).
* Add AsciiDoc syntax support.
* Ensure default setting for `export_options` when not available.

**v2.0.3 (07/18/2013)**

* Fix html exporting when file contains images in Sublime Text 3.
* Fix `smart_strong` and `meta_data` extention name for markdown.

**v2.0.2 (06/17/2013)**

* Update cherrypy server, add detail information if socket could not be created.
* Add support for file URIs in images path.

**v2.0.1 (05/12/2013)**

* Strip YAML frontmatter for Markdown files automatically.
* Check syntax name as well as filename extension for MediaWiki files.

**v2.0 (03/31/2013)**

* Added support for [Org Mode](http://orgmode.org) (Requires ruby, and gem
  `org-ruby` should be installed).
* Added support for [MediaWiki](http://www.mediawiki.org/) (Requires ruby, as
  well as gem `wikicloth`).
* Added support for [AsciiDoc](http://www.methods.co.nz/asciidoc/) (Requires ruby,
  as well as gem `asciidoctor`).
* Reviving view (redirecting to the new location) automatically after reconnected.
* Prevent Package Control for Sublime Text 3 installing this package as
  `.sublime-package` (zip archive).
* Fixed broken `ignored_renderer` setting.
* Improved Sublime Text 3 compatibility.

**v1.20 (03/15/2013)**

* Add support Sublime Text 3 (Experimental).
* Add new context command `Copy Markup as HTML`.
* Remove unused command `Sweep Cache (Remove Unused)`.
* Auto scroll now works correctly for documents contain images and MathJax equations.

**v1.12 (03/13/2013)**

* Renderes are now loaded asynchronously on startup (faster Sublime Text 2 startup).
* Add litcoffee support.


License
-------

This plugin released under MIT License:

    Copyright © 2012 - 2015 Timon Wong

    Permission is hereby granted, free of charge, to any person obtaining a copy of
    this software and associated documentation files (the "Software"), to deal in
    the Software without restriction, including without limitation the rights to
    use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
    of the Software, and to permit persons to whom the Software is furnished to do
    so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in all
    copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
    SOFTWARE.

Donors
------

[DONORS.md](./DONORS.md)


CHANGELOG.md
OmniMarkupPreviewer Changes
---------------------------

**v3.0.0 (02/28/2015)**

* Use new flat github style (old style renamed to github-v1).
* Upgrade pygments library to v2.0.2, with more lexers.
* Upgrade `Python-Markdown` package to v2.4.1 final.
* Add `.mmd` file extension support (MultiMarkdown).
* Upgrade jQuery to 2.1.3 (Say good bye to IE8).
* `server_host` setting is now taken into account when launching preview in browser.

**v2.0.9 (08/02/2014)**

* Add syntax highlighting support for reStructureText.
* Prefer to use `xdg-open` under linux environments.

**v2.0.8 (04/15/2014)**

* Fix Sublime Text 2 compatibility (crash).

**v2.0.7 (04/14/2014)**

* Fix unreliable default setting overriding, which confuses a lot of users when
  customizing their settings.
* Prevent preview from the console, which will simply lead to crash.

**v2.0.6 (01/27/2014)**

* Fix Sublime Text 2 compatibility (markdown).

**v2.0.5 (01/24/2014)**

* MathJax library is now bundled directly instead of on demand downloading.
* Python-Markdown module now upgraded to v2.4.
* Fix incompatibility when using Sublime Text 3 under XFCE.

**v2.0.4 (08/10/2013)**

* Fix ruby gems loading in OSX (Required by RDoc, AsciiDoc, Org Mode and MediaWiki).
* Add AsciiDoc syntax support.
* Ensure default setting for `export_options` when not available.

**v2.0.3 (07/18/2013)**

* Fix html exporting when file contains images in Sublime Text 3.
* Fix `smart_strong` and `meta_data` extention name for markdown.

**v2.0.2 (06/17/2013)**

* Update cherrypy server, add detail information if socket could not be created.
* Add support for file URIs in images path.

**v2.0.1 (05/12/2013)**

* Strip YAML frontmatter for Markdown files automatically.
* Check syntax name as well as filename extension for MediaWiki files.

**v2.0 (03/31/2013)**

* Added support for [Org Mode](http://orgmode.org) (Requires ruby, and gem
  `org-ruby` should be installed).
* Added support for [MediaWiki](http://www.mediawiki.org/) (Requires ruby, as
  well as gem `wikicloth`).
* Added support for [AsciiDoc](http://www.methods.co.nz/asciidoc/) (Requires ruby,
  as well as gem `asciidoctor`).
* Reviving view (redirecting to the new location) automatically after reconnected.
* Prevent Package Control for Sublime Text 3 installing this package as
  `.sublime-package` (zip archive).
* Fixed broken `ignored_renderer` setting.
* Improved Sublime Text 3 compatibility.

**v1.20 (03/15/2013)**

* Add support Sublime Text 3 (Experimental).
* Add new context command `Copy Markup as HTML`.
* Remove unused command `Sweep Cache (Remove Unused)`.
* Auto scroll now works correctly for documents contain images and MathJax equations.

**v1.12 (03/13/2013)**

* Renderes are now loaded asynchronously on startup (faster Sublime Text 2 startup).
* Add litcoffee support.

**v1.11 (12/24/2012)**

* Fix incorrect auto-scrolling behavior while pages contain images or mathjax equations.

**v1.10 (11/22/2012)**

* Fix `UnicodeEncodeError` exception while image file path conatins non-ascii characters.
* Fix missing background image for `hr` element from exported htmls.

**v1.9 (11/12/2012)**

* Provide support for exporting result to html file, images on disk will be inlined (data-url).
  You can customize the settings of the exporter through the `"export_options"` option.
* Fix incorrect code block detecting (due to wrong tab length setting) in
  markdown renderer.

**v1.8 (11/10/2012)**

* OmniMarkupPreview is now able to use user defined browser command for launching
  web browser, through the `"browser_command"` option.
* Unopened file can be previewed correctly in browser without returning "404" error.
* Update cherrypy module in order to fix a random server crash on startup in
  Windows (Refer to [CherrPy #1016])

[CherrPy #1016]: https://bitbucket.org/cherrypy/cherrypy/issue/1016/windowserror-error-6-the-handle-is-invalid

**v1.7 (11/07/2012)**

* Add option `"server_host"` for server listening address.
* Now OmniMarkupPreviewer doesn't require restart on some settings change anymore.
* On demand downloader for mathjax should work under Linux now (using `wget` or `curl`).
* Unsaved textile documents can now be previewed as well.

**v1.6 (11/03/2012)**

* [MathJax] support is now added (through the `"mathjax_enabled"` option), you can
  use `$..$` and `\(..\)` delimiters for inline math, `$$..$$` and `\[..\]` delimiters
  for display math. [MathJax] libraries will be downloaded on demand. for more
  information, visit [my blog](http://theo.im/blog/2012/11/03/latex-support-in-omnimarkuppreviewer/).
* Add support to custom the behavior of markdown renderer (through the
  `"renderer_options-MarkdownRenderer"` option).
* Responsive width on browser width change.
* Much better http server performance (Thanks to the [CherryPy] project).

[MathJax]: http://www.mathjax.org
[CherryPy]: http://www.cherrypy.org

**v1.5.1 (11/01/2012)**

* Fixed a bug that all renderers wouldn't work if any renderer raised exception
  while loading.

**v1.5 (10/31/2012)**

* Images on local machine can now be displayed corectlly.
* New `"ajax_polling_interval"` option.
* Allow users to use their own templates.

**v1.4 (10/28/2012)**

* Code blocks from [GitHub flavored markdown] is supported now, through the
  [Fenced Code Blocks Extension].
* [PHP Markdown Tables] support is added, through the [Tables Extension].
* Support strikeout extension syntax (Pandoc and GitHub) for markdown: `This ~~is deleted text.~~`
* Added `"ignored_renderers"` option to settings, in order to ignore specific
  markup renderers.

[GitHub flavored markdown]: http://github.github.com/github-flavored-markdown/
[Fenced Code Blocks Extension]: http://packages.python.org/Markdown/extensions/fenced_code_blocks.html
[PHP Markdown Tables]: http://michelf.ca/projects/php-markdown/extra/#table
[Tables Extension]: http://packages.python.org/Markdown/extensions/tables.html

**v1.3 (10/27/2012)**


OmniMarkupLib/Downloader.py
import os
import re
import subprocess
import tempfile

from .Common import PY3K

if PY3K:
    from http.client import HTTPException
    from urllib.request import ProxyHandler, HTTPRedirectHandler, build_opener, Request
    from urllib.error import HTTPError, URLError
else:
    from httplib import HTTPException
    from urllib2 import ProxyHandler, HTTPRedirectHandler, build_opener, Request
    from urllib2 import HTTPError, URLError

from . import log

# HACK: Exception about `LookupError: unknown encoding: idna`
#   will be thrown occasionally.
try:
    exec('from encodings import idna')
except:
    pass


class BinaryNotFoundError(Exception):
    pass


class NonCleanExitError(Exception):
    def __init__(self, returncode):
        self.returncode = returncode

    def __str__(self):
        return repr(self.returncode)


class CliDownloader(object):
    def __init__(self, setting):
        self.setting = setting

    def find_binary(self, name):
        for dir in os.environ['PATH'].split(os.pathsep):
            path = os.path.join(dir, name)
            if os.path.exists(path):
                return path

        raise BinaryNotFoundError('The binary %s could not be located' % name)

    def execute(self, args):
        proc = subprocess.Popen(args,
                                stdin=subprocess.PIPE,
                                stdout=subprocess.PIPE,
                                stderr=subprocess.PIPE)

        output = proc.stdout.read()
        self.stderr = proc.stderr.read()
        returncode = proc.wait()
        if returncode != 0:
            error = NonCleanExitError(returncode)
            error.output = self.stderr
            raise error
        return output


class WgetDownloader(CliDownloader):
    def __init__(self, setting):
        self.setting = setting
        self.wget = self.find_binary('wget')

    def clean_tmp_file(self):
        os.remove(self.tmp_file)

    def download(self, url, error_message, timeout, tries):
        if not self.wget:
            return False

        self.tmp_file = tempfile.NamedTemporaryFile().name
        command = [self.wget,
                   '--connect-timeout=' + str(int(timeout)),
                   '-o', self.tmp_file,
                   '-O', '-', '-U', 'OmniMarkup Downloader']

        command.append(url)

        if self.setting.http_proxy:
            os.putenv('http_proxy', self.setting.http_proxy)
            if not self.setting.https_proxy:
                os.putenv('https_proxy', self.setting.http_proxy)
        if self.setting.https_proxy:
            os.putenv('https_proxy', self.setting.https_proxy)

        while tries > 0:
            tries -= 1
            try:
                result = self.execute(command)
                self.clean_tmp_file()
                return result
            except NonCleanExitError as e:
                error_line = ''
                with open(self.tmp_file) as f:
                    for line in list(f):
                        if re.search('ERROR[: ]|failed: ', line):
                            error_line = line
                            break

                if e.returncode == 8:
                    regex = re.compile('^.*ERROR (\d+):.*', re.S)
                    if re.sub(regex, '\\1', error_line) == '503':
                        # GitHub and BitBucket seem to rate limit via 503
                        log.warning('Downloading %s was rate limited, trying again', url)
                        continue
                    error_string = 'HTTP error ' + re.sub('^.*? ERROR ', '', error_line)

                elif e.returncode == 4:
                    error_string = re.sub('^.*?failed: ', '', error_line)
                    # GitHub and BitBucket seem to time out a lot
                    if error_string.find('timed out') != -1:
                        log.warning('Downloading %s timed out, trying again', url)
                        continue

                else:
                    error_string = re.sub('^.*?(ERROR[: ]|failed: )', '\\1', error_line)

                error_string = re.sub('\\.?\s*\n\s*$', '', error_string)
                log.warning('%s %s downloading %s.', error_message, error_string, url)
            self.clean_tmp_file()
            break
        return False


class CurlDownloader(CliDownloader):
    def __init__(self, setting):
        self.setting = setting
        self.curl = self.find_binary('curl')

    def download(self, url, error_message, timeout, tries):
        if not self.curl:
            return False
        command = [self.curl,
                   '-f', '--user-agent', 'OmniMarkup Downloader',
                   '--connect-timeout', str(int(timeout)), '-sS']

        command.append(url)

        if self.setting.http_proxy:
            os.putenv('http_proxy', self.setting.http_proxy)
            if not self.setting.https_proxy:
                os.putenv('HTTPS_PROXY', self.setting.http_proxy)
        if self.setting.https_proxy:
            os.putenv('HTTPS_PROXY', self.setting.https_proxy)

        while tries > 0:
            tries -= 1
            try:
                return self.execute(command)
            except (NonCleanExitError) as e:
                if e.returncode == 22:
                    code = re.sub('^.*?(\d+)\s*$', '\\1', e.output)
                    if code == '503':
                        # GitHub and BitBucket seem to rate limit via 503
                        log.warning('Downloading %s was rate limited, trying again', url)
                        continue
                    error_string = 'HTTP error ' + code
                elif e.returncode == 6:
                    error_string = 'URL error host not found'
                elif e.returncode == 28:
                    # GitHub and BitBucket seem to time out a lot
                    log.warning('Downloading %s timed out, trying again', url)
                    continue
                else:
                    error_string = e.output.rstrip()

                log.warning('%s %s downloading %s.', error_message, error_string, url)
            break
        return False


class UrlLib2Downloader(object):
    def __init__(self, setting):
        self.setting = setting

    def download(self, url, error_message, timeout, tries):
        http_proxy = self.setting.http_proxy
        https_proxy = self.setting.https_proxy
        if http_proxy or https_proxy:
            proxies = {}
            if http_proxy:
                proxies['http'] = http_proxy
                if not https_proxy:
                    proxies['https'] = http_proxy
            if https_proxy:
                proxies['https'] = https_proxy
            proxy_handler = ProxyHandler(proxies)
        else:
            proxy_handler = ProxyHandler()
        handlers = [proxy_handler, HTTPRedirectHandler()]

        # secure_url_match = re.match('^https://([^/]+)', url)
        # if secure_url_match != None:
        #   secure_domain = secure_url_match.group(1)
        #   bundle_path = self.check_certs(secure_domain, timeout)
        #   if not bundle_path:
        #       return False
        #   handlers.append(VerifiedHTTPSHandler(ca_certs=bundle_path))
        opener = build_opener(*handlers)

        while tries > 0:
            tries -= 1
            try:
                request = Request(url, headers={"User-Agent": "OmniMarkup Downloader"})
                http_file = opener.open(request, timeout=timeout)
                return http_file.read()

            except HTTPException as e:
                log.warning('%s HTTP exception %s (%s) downloading %s.',
                            error_message, e.__class__.__name__, str(e), url)

            except HTTPError as e:
                # Bitbucket and Github ratelimit using 503 a decent amount
                if str(e.code) == '503':
                    log.warning('Downloading %s was rate limited, trying again', url)
                    continue
                log.warning('%s HTTP error %s downloading %s.',
                            error_message, str(e.code), url)

            except URLError as e:
                # Bitbucket and Github timeout a decent amount
                if str(e.reason) == 'The read operation timed out' or \
                        str(e.reason) == 'timed out':
                    log.warning('Downloading %s timed out, trying again', url)
                    continue
                log.warning('%s URL error %s downloading %s.',
                            error_message, str(e.reason), url)
            break
        return False
